cmd/kubectl/kubectl.go

```diff
@@ -119,34 +118,32 @@ func NewK0sKubectlCmd() *cobra.Command {
 	return cmd
 }
 
-func fallbackToK0sKubeconfig(args []string) error {
-	for _, arg := range args {
-		if arg == "--" {
-			// no more options
-			break
-		}
-		if arg == "--kubeconfig" || strings.HasPrefix(arg, "--kubeconfig=") {
-			// kubeconfig set via args, no need to check if k0s kubeconfig is readable
-			return nil
-		}
+func fallbackToK0sKubeconfig(cmd *cobra.Command) error {
+	kubeconfigFlag := cmd.Flags().Lookup("kubeconfig")
+	if kubeconfigFlag == nil {
+		return fmt.Errorf("kubeconfig flag not found")
 	}
 
-	if _, envSet := os.LookupEnv("KUBECONFIG"); envSet {
-		// kubeconfig environment variable set, don't override
+	if kubeconfigFlag.Changed {
+		// prioritize flag over env
+		_ = os.Unsetenv("KUBECONFIG")
 		return nil
 	}
 
-	kubeconfig := config.GetCmdOpts().K0sVars.AdminKubeConfigPath
-	// verify that k0s's kubeconfig is readable before pushing it to the env
-	file, err := os.Open(kubeconfig)
-	if err != nil {
-		return fmt.Errorf("cannot read k0s kubeconfig, is the server running? (%w)", err)
+	kubeconfig, envSet := os.LookupEnv("KUBECONFIG")
+	if envSet {
+		_ = os.Unsetenv("KUBECONFIG")
+	} else {
+		kubeconfig = config.GetCmdOpts().K0sVars.AdminKubeConfigPath
 	}
-	file.Close()
 
-	if err := os.Setenv("KUBECONFIG", kubeconfig); err != nil {
-		return fmt.Errorf("failed to set k0s kubeconfig as default: %w", err)
+	// verify that k0s's kubeconfig is readable before pushing it to the env
+	if _, err := os.Stat(kubeconfig); err != nil {
+		return fmt.Errorf("cannot stat k0s kubeconfig, is the server running?: %w", err)
```

Comment:
The file existence check was originally only performed when the k0s admin kubeconfig was used. Now it's also used for the environment variable. The env variable can actually hold a [list of kubeconfig files](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#the-kubeconfig-environment-variable), which gets merged. Even if there's not a single existing file in that list, it will still fallback to the default config (localhost:6443), IIIRC.

I'd only do the extra existence check if we're setting the k0s default.